src/hotspot/share/utilities/vmError.cpp

```diff
@@ -1629,27 +1628,42 @@ void VMError::report_and_die(int id, const char* message, const char* detail_fmt
         // Watcherthread is about to call os::die. Lets just wait.
         os::infinite_sleep();
       } else {
-        // Crash or assert during error reporting. Lets continue reporting with the next step.
-        stringStream ss(buffer, sizeof(buffer));
+        // A secondary error happened. Print brief information, but take care, since crashing
+        // here would just recurse endlessly.
+        // Any information (signal, context, siginfo etc) printed here should use the function
+        // arguments, not the information stored in *this, since those describe the primary crash.
+        char tmp[256]; // cannot use global scratch buffer
         // Note: this string does get parsed by a number of jtreg tests,
         // see hotspot/jtreg/runtime/ErrorHandling.
-        ss.print("[error occurred during error reporting (%s), id 0x%x",
+        st->print("[error occurred during error reporting (%s), id 0x%x",
                    _current_step_info, id);
-        char signal_name[64];
-        if (os::exception_name(id, signal_name, sizeof(signal_name))) {
-          ss.print(", %s (0x%x) at pc=" PTR_FORMAT, signal_name, id, p2i(pc));
+        if (os::exception_name(id, tmp, sizeof(tmp))) {
+          st->print(", %s (0x%x) at pc=" PTR_FORMAT, tmp, id, p2i(pc));
```

Comment:
Not really relevant for this PR. But do not like the inconsistency that the id is printed as hex here, decimal in os::print_siginfo for posix, and hex in os::print_siginfo for windows.

Reply:
I think the whole concept of VMError::id is not that good. It holds either SEH codes (windows), signal numbers or some enum we ourselves define. So it combines several orthogonal things, and the encoding is not clear. I have several half-done patches laying around cleaning it up, but never got around to finish them. I think some hierarchical structure would be better, e.g. errortype="crash|assert|oom", with auxilliary information depending on errortype attached. The encoding can be as dense as a 32-bit value, if one wanted. But that would also sort the printing problem, since printing depends on what the id is in the first place (e.g. signals are usually printed with d, but SEH codes as 32bit hex).